### PR TITLE
Increase API generation report verbosity

### PR DESCRIPTION
**Pull request by jhowlett-scottlogic**
- Add a verbose option -v that adds more information and fire to the command line output.
- Silenced output from shell.exec
- Only show error stack with verbose output
- Update documentation to include -v option

### CODE REVIEW

File: src/generate.js [hunk 1]
```diff
@@ -75,44 +85,74 @@ function validateGenerator(generatorPath) {
...
-    if (!options.skipValidation && !(await isValidSchema(schema))) {
-      return;
+    if(!options.skipValidation) {
+      console.log("Validating schema");
+      if(!(await isValidSchema(schema))) {
+        return;
+      }
     }
 
+    numberOfDiscoveredModels = Object.keys(schema.components.schemas).length;
```
[the remark this answers is not in the document]
Comment (jhowlett-scottlogic) [reply src/generate.js:148]: needed after closure of try {} so pre-define outside block as a let on line 106

File: src/generate.js [hunk 2]
```diff
@@ -75,44 +85,74 @@ function validateGenerator(generatorPath) {
...
     if (isUrl(generatorUrlOrPath)) {
       // if the generator is specified as a git URL, clone it into a temporary directory
       const generatorUrl = generatorUrlOrPath;
       if (generatorUrl.endsWith(".git")) {
         generatorPath = temporaryFolder = fs.mkdtempSync(
           path.join(os.tmpdir(), "generator")
         );
-        console.log(`Cloning generator from ${generatorUrl}`);
-        shell.exec(`git clone ${generatorUrl} ${generatorPath}`);
+        if(options.verbose) console.log(`Creating temporary folder '${generatorPath}'`);
```
[the remark this answers is not in the document]
Comment (jhowlett-scottlogic) [reply src/generate.js:118]: It definitely can be and should be done, i will add it

File: src/generate.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (ColinEberhardt) [reply src/generate.js:148]: Good point, wasn’t obvious from the diff

File: src/generate.js [hunk 3]
```diff
@@ -76,43 +87,70 @@ function validateGenerator(generatorPath) {
...
+  log.verbose("     )                             (    (      (                           ");
+  log.verbose("  ( /(                      (      )\\ ) )\\ )   )\\ )                        ");
+  log.verbose("  )\\())           (         )\\    (()/((()/(  (()/(      (    (  (     (   ");
+  log.verbose(" ((_)\\   `  )    ))\\  (  ((((_)(   /(_))/(_))  /(_)) (   )(   )\\))(   ))\\  ");
+  log.verbose("   ((_)  /(/(   /((_) )\\ ))\\ _ )\\ (_)) (_))   (_))_| )\\ (()\\ ((_))\\  /((_) ");
+  log.verbose("  / _ \\ ((_)_\\ (_))  _(_/((_)_\\(_)| _ \\|_ _|  | |_  ((_) ((_) (()(_)(_))   ");
+  log.verbose(" | (_) || '_ \\)/ -_)| ' \\))/ _ \\  |  _/ | |   | __|/ _ \\| '_|/ _` | / -_)  ");
+  log.verbose("  \\___/ | .__/ \\___||_||_|/_/ \\_\\ |_|  |___|  |_|  \\___/|_|  \\__, | \\___|  ");
+  log.verbose("        |_|                                                  |___/         ");
+  log.verbose("");
```
Comment (jhowlett-scottlogic) [inline src/generate.js:101]: moved ASCII art into generate to prevent logging class from having to handle function inputs

File: src/log.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
+let logLevel = 0;
+
+const logLevels = {
+    standard: 0,
+    verbose: 1
+}
```
Comment (jhowlett-scottlogic) [inline src/log.js:6]: using this method of logging to allow for new levels to be added easily. For example a 'debug' level could be added with minimal code changes

**Comment by jhowlett-scottlogic** [comment]
I have added log.js which handles all of the logging. 

I have made some inline comments to help explain so design decisions

File: src/generate.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (jhowlett-scottlogic) [reply src/generate.js]: Added this

**Review by ColinEberhardt — APPROVED** [review]

**Comment by ColinEberhardt** [comment]
Great job, thanks @jhowlett-scottlogic  👍

**Pull request merged**